src/celcompat/bit.h

```diff
@@ -0,0 +1,183 @@
+// bit.h
+//
+// Copyright (C) 2023-present, the Celestia Development Team
+//
+// This program is free software; you can redistribute it and/or
+// modify it under the terms of the GNU General Public License
+// as published by the Free Software Foundation; either version 2
+// of the License, or (at your option) any later version.
+
+#pragma once
+
+#if __cplusplus >= 202002L && __has_include(<bit>)
+#include <bit>
+#endif
+
+#include <cstdint>
+#include <type_traits>
+
+#if defined(__has_builtin)
+#  if __has_builtin(__builtin_bit_cast)
+#    define HAVE_BUILTIN_BIT_CAST 1
+#  endif
+#elif defined(__clang__)
+#  if __clang_major__ >= 9
+#    define HAVE_BUILTIN_BIT_CAST 1
+#  endif
+#  define HAVE_BUILTIN_MEMCPY 1
+#elif defined(__GNUC__)
+#  if __GNUC__ >= 11
+#    define HAVE_BUILTIN_BIT_CAST 1
+#  endif
+#  define HAVE_BUILTIN_MEMCPY 1
+#endif
+
+#if !defined(HAVE_BUILTIN_BIT_CAST) && !defined(HAVE_BUILTIN_MEMCPY)
+#include <cstring> // memcpy
+#endif
+
+/**
+ * std::endian implementation
+ */
+namespace celestia::compat
+{
+
+#if __cpp_lib_endian
+
+using std::endian;
+
+#else
+
+enum class endian
+{
+#if defined(_MSC_VER) && !defined(__clang__)
+    little = 1234,
+    big    = 4321,
+    native = little
+#else
+    little = __ORDER_LITTLE_ENDIAN__,
+    big    = __ORDER_BIG_ENDIAN__,
+    native = __BYTE_ORDER__
+#endif
+};
+
+#endif // __cpp_lib_endian
+
+} // namespace celestia::compat
+
+/**
+ *  std::bit_cast implementation
+ */
+namespace celestia::compat
+{
+
+#if __cpp_lib_bit_cast
+
+using std::bit_cast;
+
+#else
+
+template<typename To, typename From,
+         std::enable_if_t<std::conjunction_v<std::bool_constant<sizeof(To) == sizeof(From)>,
+                                             std::is_trivially_copyable<To>,
+                                             std::is_trivially_copyable<From>>,
+                          bool> = false>
+[[nodiscard]]
+#if defined(HAVE_BUILTIN_BIT_CAST) || defined(HAVE_BUILTIN_MEMCPY)
+constexpr
+#else
+inline
+#endif
+To bit_cast(const From src) noexcept
+{
+#ifdef HAVE_BUILTIN_BIT_CAST
+    return __builtin_bit_cast(To, src);
+#else
+    static_assert(std::is_trivially_constructible_v<To>,
+        "This implementation additionally requires "
+        "destination type to be trivially constructible");
+
+    To dst;
+#ifdef HAVE_BUILTIN_MEMCPY
+    __builtin_memcpy(&dst, &src, sizeof(To));
+#else
+    std::memcpy(&dst, &src, sizeof(To));
+#endif
+    return dst;
+#endif
+}
+
+#endif // __cpp_lib_bit_cast
+
+} // namespace celestia::compat
+
+/**
+ * std::byteswap implementation
+ */
+namespace celestia::compat
+{
+
+#if __cpp_lib_byteswap
+
+using std::byteswap;
+
+#else
+
+namespace
```

Comment:
This should be named (e.g. "internal" or "impl"), otherwise it will create a new anonymous namespace in all translation units.